File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/queues/providers/RepeatingPositionedFileQueue.java
package com.lasthopesoftware.bluewater.client.playback.queues.providers;

import com.lasthopesoftware.bluewater.client.playback.file.PositionedFile;
import com.lasthopesoftware.bluewater.client.playback.queues.IPositionedFileQueue;

import java.util.ArrayDeque;
import java.util.List;
import java.util.Queue;

final class RepeatingPositionedFileQueue implements IPositionedFileQueue {
	private final Queue<PositionedFile> playlist;

	RepeatingPositionedFileQueue(List<PositionedFile> playlist) {
		this.playlist = new ArrayDeque<>(playlist);
	}

	@Override
	public PositionedFile poll() {
		if (playlist.size() == 0)
			return null;

		final PositionedFile positionedFile = playlist.poll();
		playlist.offer(positionedFile);

		return positionedFile;
	}

	@Override
	public PositionedFile peek() {
		return playlist.size() > 0 ? playlist.peek() : null;
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/preparation/queues/RepeatingPositionedFileQueue.kt
package com.lasthopesoftware.bluewater.client.playback.file.preparation.queues

import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.playback.file.PositionedFile
import java.util.ArrayDeque

internal class RepeatingPositionedFileQueue(override val libraryId: LibraryId, playlist: List<PositionedFile>) :
    PositionedFileQueue {
    private val playlist by lazy { ArrayDeque(playlist) }

    override fun poll(): PositionedFile? {
        val positionedFile = playlist.poll() ?: return null
        playlist.offer(positionedFile)
        return positionedFile
    }

    override fun peek(): PositionedFile? = playlist.peek()
}
